### ISSUE TITLE
[s2MOP /] Shortcode

### ISSUE THREAD

**Issue by jaswsinc** [websharks/s2member-kb#131]
---
##### KB Article Creation Checklist

- [x] Write initial draft for this KB Article; label this issue `draft` and either `questions` or `tutorials`
- [x] Add [required YAML configuration](https://github.com/websharks/team/wiki/KB-Articles-::-Required-YAML-Configuration)
- [x] Add Tags for this KB Article to the YAML config (see [YAML Keys (Explained)](https://github.com/websharks/team/wiki/KB-Articles-::-Required-YAML-Configuration#yaml-keys-explained))
- [x] Edit and finalize draft for publishing (remove `draft` label, add `draft-finalized` label)
- [x] Assign Issue to yourself and [create Markdown file](https://github.com/websharks/team/wiki/KB-Articles-::-Creating-the-Markdown-File) (remove `draft-finalized` label, add `pending`)
- [x] Project Lead: Review and Publish KB Article (remove `pending` label, add `published` label)

##### Additional TODOs

_Additional TODO list items go here._

---

**Markdown File:** http://wsharks.com/18ZG8VX
**Published KB Article:** http://s2member.com/kb-article/s2mop-shortcode/

---

**Comment by Lifeischoice** [comment]
Is there a way to redirect this to a different page?

**Comment by jaswsinc** [comment]
@Lifeischoice Redirect the Membership Options Page to somewhere else conditionally?

**Comment by Lifeischoice** [comment]
Yes...thanks

**Comment by jaswsinc** [comment]
@Lifeischoice You could do that, yes. It would need to be done with JavaScript though, because a shortcode is dealing with output, not HTTP headers. So you could do it, but it might cause a flash of the Membership Options Page first, followed by an immediate redirect.

```html
<script>
    location.href = 'http://example/redirect';
</script>
```

Keep in mind that the `[s2MOP /]` shortcode is there to help you _avoid_ the need for redirections. Using the `[s2MOP /]` shortcode, you should be able to adjust the display of a single Membership Options Page based on what type of access was being attempted. So my advice would be _not_ to use redirections, but to take full advantage of the dynamic output that you can achieve using `[s2MOP /]` and perhaps `[s2If /]` also. See: [\[s2If /\] Simple Shortcode Conditionals](http://s2member.com/kb-article/s2if-simple-shortcode-conditionals/)

i.e., One page that is dynamic, is usually easier than 12 static pages that are not.

**Comment by raamdev** [comment]
@jaswsinc @Lifeischoice Referencing this related s2Member Feature Request: https://github.com/websharks/s2member/issues/595

**Comment by Lifeischoice** [comment]
@jaswsinc Thanks but I am not sure I follow. What I would like is to be able to use the [s2Mop /] shortcode in a way that can redirect to a designated page for a specific offer. For example, the default  Membership Options page has 3 offers - a, b, c. However, for offer b, I have a sales page with more specific information for that offer b. Currently, if someone attempts to access the pages protected by offer b, the user is redirected to the default Membership Options page with the 3 offers. What I would prefer is for the user to be redirected to the page with the specific "b" sales page. If a user attempts to access pages for a and c, they may be redirected to the default membership options page.

**Comment by jaswsinc** [comment]
@Lifeischoice You might want to take a look at [WP Snippets](https://wordpress.org/plugins/wp-snippets/) and use that in concert with `[s2MOP /]`

### `[s2MOP /]` + `[snippet /]` Example

```
[s2MOP required_type="ccap" required_value="pro"]
    [snippet slug="pro_offer_content" /]
[/s2MOP]
```

So here I am detecting that a visitor is attempting to access something that requires CCAP `pro`, and I'm pulling the content for the pro offer into the Membership Options Page dynamically.

---

Or, if you really do want to go the redirect route instead, you could use a PHP filter that scans the s2Member MOP Vars whenever a visitor lands on the Membership Options Page, and redirect them away, _before_ the content loads; i.e., do the redirection in PHP, not with a shortcode. For instance, see [this section of the article about MOP Vars](http://s2member.com/kb-article/s2mop-shortcode/#toc-dc55bf9a), and work to integrate that into a redirection routine via PHP.

### `wp_redirect()` Example (in PHP code)

Create this directory and file:
`/wp-content/mu-plugins/s2-hacks.php`

```php
<?php
add_action('template_redirect', function () {
    if (empty($_REQUEST['_s2member_vars'])) {
        return; // Nothing to do here.
    }
    if (!is_page(S2MEMBER_MEMBERSHIP_OPTIONS_PAGE_ID)) {
        return; // Nothing to do here.
    }
    @list($restriction_type, $requirement_type, $requirement_type_value, $seeking_type, $seeking_type_value, $seeking_uri)
        = explode('..', stripslashes((string) $_REQUEST['_s2member_vars']));

    if ($requirement_type === 'ccap' && $requirement_type_value === 'pro') {
        wp_redirect(home_url('/pro-offer/'));
        exit;
    }
});
```

**Comment by jaswsinc** [comment]
Referencing a similar request here: https://github.com/websharks/s2member/issues/784#issue-116413246

**Comment by Lifeischoice** [comment]
@jaswsinc...have tried both methods with no success. Fit the first, installed/activated wp snippet and added your snippet. Then added snippet shotcode to the top of mop.
The second way was added to s2-hack. Neither worked for me...so if you dont mind adding more specific instructions for newbie...much appreciated.

**Comment by KTS915** [comment]
@jaswsinc, @Lifeischoice,

I have been playing around with this to get a different redirect going. Adapting what I have done to this, I suggest you try using the following code in your ````/wp-content/mu-plugins/s2-hacks.php```` file:

````<?php
add_action('template_redirect', function () {
    if (!is_page(S2MEMBER_MEMBERSHIP_OPTIONS_PAGE_ID)) {
        return; // Nothing to do here.
    }
    else if (!empty($_REQUEST['_s2member_vars'])) 
		@list($restriction_type, $requirement_type, $requirement_type_value, $seeking_type, $seeking_type_value, $seeking_uri)
        = explode('..', stripslashes((string) $_REQUEST['_s2member_vars']));

    if (!empty($requirement_type) && $requirement_type === 'ccap') {
        if (!empty($requirement_type_value) && $requirement_type_value === 'pro') {
            wp_redirect(esc_url(home_url('/pro-offer/')));
            exit;
        }
    }
});````

**Comment by Lifeischoice** [comment]
@KTS915 
Thanks, however, it still did not redirect for me. This is the step I followed.
1. at the top of my MOP page, I added  [s2MOP] You were trying to access content that requires a %%RESTRICTION_TYPE%% that you don't have.[/s2MOP]
2. protected my content page with level 1 and ccap "pro"
3. added the snippet above to mu-plugins/s2-hacks.php
4. logged out
5. I then visited the url of page protected with ccap = pro and was redirected to MOP (not to pro-offer page)

**Comment by KTS915** [comment]
@Lifeischoice,

That's not how to use my code. You don't want to add anything to the MOP page at all. 

Just create a file called ````s2-hacks.php``` and paste the code into it. Then upload the file into the ````wp-content/mu-plugins```` folder. And if that folder does not already exist, create it.

**Comment by Lifeischoice** [comment]
@KTS915 Thanks..I have a mu-plugins folder already with s2-hacks.php and I have deleted step 1 above and it still did not work..the only difference is in step 5 with no display of the reason for redirect.  Not sure what I am doing incorrectly. Can you confirm that the rest of the steps I outlined is ok (steps 2-5)?

**Comment by KTS915** [comment]
@Lifeischoice,

Yes, I set out the whole procedure because I know from experience that others read these threads much later and then ask how to make them work for them, and it's hard to remember!

Sorry, though, that it's not working for you. And I hope you won't be offended if I ask whether you have enabled MOP Vars.

**Comment by Lifeischoice** [comment]
@KTS915 
Offended? Not at all...I am glad you are helping me - Much appreciated..Yes, I have enabled MOP Vars under S2member (pro) -> Membership Option Page admin menu.

**Comment by KTS915** [comment]
OK, good! Is your Pro Offer page generally accessible, or have you protected it in some way?

**Comment by Lifeischoice** [comment]
thanks...the Pro Offer page is not protected.

**Comment by KTS915** [comment]
Well, I must admit I'm baffled then. You seem to have done everything I can think of. I'm afraid that you're going to need the much greater expertise of @jaswsinc or @raamdev!

**Comment by Lifeischoice** [comment]
I finally got it to work...my error was that I was protecting with both page restriction and ccap! I removed the page restriction and voila! it worked. a big that you to @KTS915, @jaswsinc . I am very  grateful.

**Comment by KTS915** [comment]
Oh, great! Thanks for letting us know!

**Comment by nigel7119** [comment]
Is there a way to redirect to the login page or include the login on the MOP page and then open the content?  We have a private Facebook group and want to share posts from our site to there to promote them.  These users will have restricted access to both the Facebook group and the site.  The problem is when I share a restricted post to Facebook the link automatically redirects me to the MOP page but I need users to be able to login and see the post as they will already have the correct registration...they just need to login.

**Comment by KTS915** [comment]
@nigel7119, see the code sample I've provided in the websharks forum.

**Comment by ghost** [comment]
hi, when I use any of the s2MOP shortcodes, they display on my webpage instead of just the message between the square brackets - please help @raamdev @jaswrks Thanks

**Comment by ghost** [comment]
I think I figured it out - it's for s2member PRO only... isn't it?
